DMCompiler/DM/Visitors/DMExpressionBuilder.cs

```diff
@@ -554,156 +532,138 @@ public enum ScopeMode {
                                 return expr;
                             }
 
+                            if (property == null) {
+                                throw new UnknownIdentifierException(deref.Location, field);
+                            }
+
+                            if ((property.ValType & DMValueType.Unimplemented) == DMValueType.Unimplemented) {
+                                DMCompiler.UnimplementedWarning(deref.Location,
+                                    $"{prevPath}.{field} is not implemented and will have unexpected behavior");
+                            }
+
                             operations = new Dereference.Operation[newOperationCount];
                             astOperationOffset = i + 1;
                             i = -1;
+                            prevPath = property.Type;
+                            pathIsFuzzy = prevPath == null;
+                            continue;
                         }
-                    }
 
-                    if (property == null) {
-                        throw new UnknownIdentifierException(deref.Location, field);
+                        if (property == null) {
+                            throw new UnknownIdentifierException(deref.Location, field);
+                        }
                     }
 
-                    if ((property.ValType & DMValueType.Unimplemented) == DMValueType.Unimplemented) {
-                        DMCompiler.UnimplementedWarning(deref.Location, $"{prevPath}.{field} is not implemented and will have unexpected behavior");
-                    }
+                    operation = new Dereference.FieldOperation {
+                        Safe = fieldOperation.Safe,
+                        Identifier = fieldOperation.Identifier,
+                        Path = property?.Type
+                    };
 
-                    prevPath = property.Type;
-                    pathIsFuzzy = false;
-                }
-                    break;
-
-                case DMASTDereference.OperationKind.FieldSearch:
-                case DMASTDereference.OperationKind.FieldSafeSearch:
-                    // TODO: im pretty sure types should be inferred if a field with their name only exists in a single place, sounds cursed though
-                    operation.Identifier = astOperation.Identifier.Identifier;
-                    operation.GlobalId = null;
-                    operation.Path = null;
-                    prevPath = null;
-                    pathIsFuzzy = true;
+                    prevPath = property?.Type;
+                    pathIsFuzzy = property == null;
                     break;
+                }
 
-                case DMASTDereference.OperationKind.Index:
-                case DMASTDereference.OperationKind.IndexSafe:
+                case DMASTDereference.IndexOperation indexOperation:
                     // Passing the path here is cursed, but one of the tests seems to suggest we want that?
```

Comment:
This comment was on the `DMExpression.Create()` call, move it above the `Index` assignment for clarity